x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/analytics/process/AnalyticsResultProcessor.java

```diff
@@ -112,4 +114,14 @@ private void joinCurrentResults() {
             }
         }
     }
+
+    private void checkIdHashesMatch(DataFrameDataExtractor.Row row, AnalyticsResult result) {
+        if (row.getIdHash().equals(result.getIdHash()) == false) {
+            StringBuilder msg = new StringBuilder();
+            msg.append("Detected id hash mismatch for document with id [" + row.getHit().getId() + "]; ");
+            msg.append("expected hash was [" + row.getIdHash() + "] but result had [" + result.getIdHash() + "]; ");
+            msg.append("this implies that the data frame index copy has been modified externally while analysis was running");
```

Comment:
How is verifying that the hash of the IDs the same as verifying that data was not mutated? It seems to me that all this is verifying is that the doc Ids are unchanged.

Reply:
The intention is not to verify the data was not mutated. The reason we add this is to sanity check the join of the results with the dataframe rows is correct. We depend on the data in the dataframe copy index to be immutable and the fact that the c++ should write out results in the same order as the input it received.

Reply:
Then what is meant by "implies that the data frame index copy has been modified externally"?

Reply:
I open to better suggestions here! The intention is to communicate that when this appears it means that most probably the user modified the copy of the data frame that is analysed while the analysis was running. The other scenario is of course that there's a bug on our side.

Reply:
How about "this implies the data frame index was modified while the analysis was running. We rely on this index being immutable during a running analysis and so the results will be unreliable." Also maybe something about how to recover?

Reply:
From what I am reading, all this really implies is that the ID of a document has changed. I suppose I am reading this as an `iff` which would mean that we are also accounting for data changing, which we are not, as the doc could be updated just fine by the user but the ID is the same.

Reply:
Or more importantly new rows added. This is perhaps the most likely scenario. We also rely on the documents being visited in a consistent order, which this checks. But to your point, we could easily fold in a checksum the document features with the usual `^= hash(feature) + 0x9e3779b9 + (seed<<6) + (seed>>2)`.

Reply:
That is a fair point. Instead of a hash of the document id we can write a hash of the document's values. Just passing `SearchHit.hashCode()` should do the job. If we're up for this, we'll just need to also change the field name of the result from `id_hash` to something like `dataframe_row_hash`.

Reply:
As mentioned offline, I think the name `checksum` captures what we want.